File: bffs/web-bff/src/routes/itemRoutes.js
const express = require("express");
const itemGrpcClient = require("../clients/itemGrpcClient");
const rentalClient = require("../clients/rentalClient");

const router = express.Router();

router.get("/", async (_req, res, next) => {
  try {
    const items = await itemGrpcClient.getAllItems();
    return res.json(items);
  } catch (error) {
    return next(error);
  }
});

router.get("/available", async (_req, res, next) => {
  try {
    const items = await itemGrpcClient.getAvailableItems();
    return res.json(items);
  } catch (error) {
    return next(error);
  }
});

router.get("/category/:category", async (req, res, next) => {
  try {
    const items = await itemGrpcClient.getItemsByCategory(req.params.category);
    return res.json(items);
  } catch (error) {
    return next(error);
  }
});

router.get("/owner/:ownerId", async (req, res, next) => {
  try {
    const items = await itemGrpcClient.getItemsByOwner(req.params.ownerId);
    return res.json(items);
  } catch (error) {
    return next(error);
  }
});

router.get("/:id/availability", async (req, res, next) => {
  try {
    const result = await itemGrpcClient.checkItemAvailability(req.params.id);
    return res.json(result);
  } catch (error) {
    return next(error);
  }
});

router.post("/:id/rentals", async (req, res, next) => {
  try {
    const itemResult = await itemGrpcClient.getItem(req.params.id);

    if (!itemResult.found || !itemResult.item) {
      return res.status(404).json({ error: "Item not found" });
    }

    const availability = await itemGrpcClient.checkItemAvailability(req.params.id);

    if (!availability.available) {
      return res.status(409).json({ error: "Item is not available" });
    }

    const rental = await rentalClient.createRental({
      itemId: req.params.id,
      borrowerId: req.body.borrowerId,
      ownerId: itemResult.item.ownerId,
      startDate: req.body.startDate,
      endDate: req.body.endDate,
      message: req.body.message || ""
    });

    return res.status(201).json(rental);
  } catch (error) {
    return next(error);
  }
});

router.get("/:id", async (req, res, next) => {
  try {
    const result = await itemGrpcClient.getItem(req.params.id);

    if (!result.found || !result.item) {
      return res.status(404).json({ error: "Item not found" });
    }

    return res.json(result.item);
  } catch (error) {
    return next(error);
  }
});

router.post("/", async (req, res, next) => {
  try {
    const result = await itemGrpcClient.createItem(req.body);

    if (!result.success) {
      return res.status(400).json({
        error: result.message
      });
    }

    return res.status(201).json(result);
  } catch (error) {
    return next(error);
  }
});

router.put("/:id", async (req, res, next) => {
  try {
    const result = await itemGrpcClient.updateItem(req.params.id, req.body);

    if (!result.success) {
      return res.status(404).json({
        error: result.message
      });
    }

    return res.json(result);
  } catch (error) {
    return next(error);
  }
});

router.patch("/:id/status", async (req, res, next) => {
  try {
    const result = await itemGrpcClient.updateItemStatus(
      req.params.id,
      req.body.status
    );

    if (!result.success) {
      return res.status(400).json({
        error: result.message
      });
    }

    return res.json(result);
  } catch (error) {
    return next(error);
  }
});

router.delete("/:id", async (req, res, next) => {
  try {
    const result = await itemGrpcClient.deleteItem(req.params.id);

    if (!result.success) {
      return res.status(404).json({
        error: result.message
      });
    }

    return res.json(result);
  } catch (error) {
    return next(error);
  }
});

module.exports = router;

File: bffs/mobile-bff/app/routers/item_routes.py
from fastapi import APIRouter, HTTPException

from app.clients import item_client
from app.clients import rental_client
from app.schemas.mobile_schemas import CreateMobileRentalRequest

router = APIRouter(prefix="/api/mobile/items", tags=["Mobile Items"])


@router.get("/available")
async def available_items():
    items = await item_client.get_available_items()

    return [
        {
            "id": item["id"],
            "name": item["name"],
            "category": item["category"],
            "location": item["location"],
            "status": item["status"],
        }
        for item in items
    ]


@router.get("/category/{category}")
async def items_by_category(category: str):
    items = await item_client.get_items_by_category(category)

    return [
        {
            "id": item["id"],
            "name": item["name"],
            "category": item["category"],
            "location": item["location"],
            "status": item["status"],
        }
        for item in items
    ]


@router.get("/{item_id}")
async def item_details(item_id: str):
    item = await item_client.get_item(item_id)

    if item is None:
        raise HTTPException(status_code=404, detail="Item not found")

    return item


@router.post("/{item_id}/rentals")
async def create_rental_for_item(item_id: str, request: CreateMobileRentalRequest):
    item = await item_client.get_item(item_id)

    if item is None:
        raise HTTPException(status_code=404, detail="Item not found")

    available = await item_client.check_item_availability(item_id)

    if not available:
        raise HTTPException(status_code=409, detail="Item is not available")

    rental_payload = {
        "itemId": item_id,
        "borrowerId": request.borrowerId,
        "ownerId": item["ownerId"],
        "startDate": request.startDate,
        "endDate": request.endDate,
        "message": request.message,
    }

    try:
        rental = await rental_client.create_rental(rental_payload)

        return {
            "id": rental.get("id"),
            "itemId": rental.get("itemId"),
            "borrowerId": rental.get("borrowerId"),
            "ownerId": rental.get("ownerId"),
            "status": rental.get("status"),
            "startDate": rental.get("startDate"),
            "endDate": rental.get("endDate"),
        }
    except Exception as ex:
        raise HTTPException(status_code=502, detail=str(ex))